### PR TITLE
Check for pdf.js and pdf.mjs in qutebrowser/browser/pdfjs.py

### PR DESCRIPTION
**Pull request by daerich**
upstream shifted to mjs modules as of
https://github.com/mozilla/pdf.js/commit/927e50f5d48a67e76f2a51c112ea5a98867822fe

If you try using the recent pdf.js release directly from upstream as indicated in the `no_pdfs.html` render
qutebrowser fails to validate the now named "pdf.___mjs___" Javascript module.
Add some logic to rectify the situation.
This has been tested on a Fedora 39 workstation.

Can probably be reverted and replaced with a simple `sed 's/pdf.js/pdf.mjs/g'` once all distros got the hint in the "near" future.

### CODE REVIEW

File: qutebrowser/browser/pdfjs.py [hunk 1]
```diff
@@ -49,7 +49,8 @@ def generate_pdfjs_page(filename, url):
         filename: The filename of the PDF to open.
         url: The URL being opened.
     """
-    if not is_available():
+    pdfjs_bin = is_available()
```
Comment (The-Compiler) [inline qutebrowser/browser/pdfjs.py:52]:
```suggestion
    pdfjs_name = is_available()
```

or somesuch - it's not a binary after all, so that name seems a bit confusing.

File: qutebrowser/browser/pdfjs.py [hunk 2]
```diff
@@ -49,7 +49,8 @@ def generate_pdfjs_page(filename, url):
...
+    if not pdfjs_bin:
```
Comment (The-Compiler) [inline qutebrowser/browser/pdfjs.py:53]:
```suggestion
    if pdfjs_bin is None:
```

while we should never get an empty string here, I still think it's better to be explicit about `None` checks.

File: qutebrowser/browser/pdfjs.py [hunk 3]
```diff
@@ -62,7 +63,7 @@ def generate_pdfjs_page(filename, url):
                         '</body><script>{}</script>'.format(script))
     # WORKAROUND for the fact that PDF.js tries to use the Fetch API even with
     # qute:// URLs.
-    pdfjs_script = '<script src="../build/pdf.js"></script>'
+    pdfjs_script = '<script src="../build/{}"></script>'.format(pdfjs_bin)
```
Comment (The-Compiler) [inline qutebrowser/browser/pdfjs.py:66]: We usually use f-strings for new code:

```suggestion
    pdfjs_script = f'<script src="../build/{html.escape(pdfjs_bin)}"></script>'
```

I also added a [`html.escape` call](https://docs.python.org/3/library/html.html#html.escape) - while strictly speaking we shouldn't have any HTML special chars in it, it's still a footgun for the future if we don't escape the value, and it certainly won't hurt.

File: qutebrowser/browser/pdfjs.py [hunk 4]
```diff
@@ -202,15 +203,30 @@ def _read_from_system(system_path, names):
...
+    final_ext = None
+    exts = ['pdf.js', 'pdf.mjs']
+    for ext in exts:
+        try:
+            final_ext = ext
+            get_pdfjs_res('build/' + ext)
+        except PDFJSNotFound:
+            final_ext = None
+    if not final_ext:
+        raise PDFJSNotFound('build/pdf.js') from None
```
Comment (The-Compiler) [inline qutebrowser/browser/pdfjs.py:217]: I find this logic quite hard to follow. Something like

```suggestion
        try:
            get_pdfjs_res(f'build/{ext}')
            return ext
        except PDFJSNotFound:
            pass

    raise PDFJSNotFound('build/pdf.js')
```

seems to make things a bit simpler. I also removed the `from None` - you're not in an `except ...:` there, so it does nothing.

File: qutebrowser/browser/pdfjs.py [hunk 5]
```diff
@@ -202,15 +203,30 @@ def _read_from_system(system_path, names):
...
+    return final_ext
+
+
 def is_available():
```
Comment (The-Compiler) [inline qutebrowser/browser/pdfjs.py:221]: Given that this does something entirely different now (in terms of return value), this should be renamed.

However, we already have `_get_pdfjs_basename` anyways, with the difference that this function also checks for `web/viewer.html`. Wouldn't it be simpler to just leave this one as-is, and call `_get_pdfjs_basename` in `generate_pdfjs_page`?

File: tests/unit/browser/test_pdfjs.py [hunk 6]
```diff
@@ -182,7 +182,7 @@ def test_read_from_system_oserror(tmpdir, caplog, unreadable_file):
     assert message.startswith('OSError while reading PDF.js file:')
 
 
-@pytest.mark.parametrize('available', [True, False])
+@pytest.mark.parametrize('available', ['pdf.js', None])
 def test_is_available(available, mocker):
```
Comment (The-Compiler) [inline tests/unit/browser/test_pdfjs.py:186]: `is_available` is patched in various other locations in this file as well, which probably should be changed accordingly too.

File: tests/unit/browser/test_pdfjs.py
Comment (The-Compiler) [inline tests/unit/browser/test_pdfjs.py:1]: The new `_get_pdfjs_basename` function should have a test as well. Happy to take care of that one if you prefer, though.

**Review by The-Compiler — CHANGES_REQUESTED** [review]
Thanks! I added various comments, since I think a couple of things could be simplified or seem confusing to me.

**Comment by daerich** [comment]
Hi there @The-Compiler ,
thanks for taking the time to review my patch!
I already had some worries about the preferred codingstyle and how to integrate my changes nicely
into the existing codebase, so I am really grateful for your suggestions.
I'll try to integrate them around the weekend when I have a little spare time on my hands.

Regarding the unittest situation, I would greatly help me if you would try to come up with some ideas.
I haven't really gotten my feet wet with python unit testing and I think it would speed things up once
I have rewritten my patches in line with your recommendations.
Thanks again for your time!

File: qutebrowser/browser/pdfjs.py (hunk 3 above)
Comment (daerich) [reply qutebrowser/browser/pdfjs.py]: I have renamed the html local variable to viewhtml as it otherwise shadows the html module identifier

File: qutebrowser/browser/pdfjs.py (hunk 5 above)
Comment (daerich) [reply qutebrowser/browser/pdfjs.py:220]: The `is_available` function has been reverted to its original form with a commented check for `pdf.mjs` to retain
some meaningful history.
`_get_pdfjs_basename` now returns either a `str` or `None` given the availability of the mjs-module.
The branching for `no_pdfjs.html` has been adapted to include a testing expression for the function.

File: qutebrowser/browser/pdfjs.py (hunk 1 above)
Comment (daerich) [reply qutebrowser/browser/pdfjs.py]: The variable has been renamed and is more expressive now.

File: qutebrowser/browser/pdfjs.py (hunk 2 above)
Comment (daerich) [reply qutebrowser/browser/pdfjs.py]: That oversight has been fixed and it should be more pythonic

File: qutebrowser/browser/pdfjs.py (hunk 4 above)
Comment (daerich) [reply qutebrowser/browser/pdfjs.py]: duly noted and simplified

File: tests/unit/browser/test_pdfjs.py (hunk 6 above)
Comment (daerich) [reply tests/unit/browser/test_pdfjs.py]: as the function retains it original signature now, the patching has been made superfluous.

**Comment by toofar** [comment]
Had a brief test of this today (on Qt 6.5.2). Testing with `python3 -m qutebrowser -T -s content.pdfjs true :version https://www.ipcc.ch/site/assets/uploads/2021/06/Fact_sheet_AR6.pdf`

This PR does indeed make PDFjs work with the latest version (retrieved via `python3 ./scripts/dev/update_3rdparty.py`).

I haven't tested if it still works with older pdfjs versions too.

The `:version` page says: `pdf.js: no`, might be another place that needs changing.

The PDFs seem to render fine but there are a couple of JS error messages that pop up (red, scary!). This looks like its related to loading a localisation library over the fetch API. The error is:

    ERROR: JS: [qute://pdfjs/web/viewer.mjs:13562] Fetch API cannot load qute://pdfjs/web/locale/locale.json. URL scheme "qute" is not supported.

Looking at the network tab I don't see the failed request there. I don't see any other requests made via the fetch API made from a PDFjs page, so I guess this is just one new type of request with the new version.

That error message is coming from: https://github.com/qt/qtwebengine-chromium/blob/14780aaa6b0f85a6912878b3874c18e63c99a5d9/chromium/third_party/blink/renderer/core/fetch/fetch_manager.cc#L721

Which I assume comes from one of the places earlier in the same file where it checks `SchemeRegistry::ShouldTreatURLSchemeAsSupportingFetchAPI()`.

Hmmm, could/should we be adding qute:// to there? I assume if we do we will still be protected by normal CORS and CSRF.

**Comment by The-Compiler** [comment]
Thanks for having a look! See https://github.com/qutebrowser/qutebrowser/issues/7999#issuecomment-1808178828 - for now, I'd look into possibly fixing our monkeypatching to stop PDF.js from using the Fetch API.

I'd rather not weaken permissions for `qute:` wholesale, as it will affect a lot. I think we should split it into `qute-pdfjs://` etc. and set fine-grained per-scheme permissions where needed. I'm pretty sure there is an issue for this somewhere, but couldn't find it anymore last time I looked.

File: qutebrowser/browser/pdfjs.py [hunk 7]
```diff
@@ -49,24 +50,25 @@ def generate_pdfjs_page(filename, url):
...
 
     script = _generate_pdfjs_script(filename)
-    html = html.replace('</body>',
+    viewhtml = viewhtml.replace('</body>',
                         '</body><script>{}</script>'.format(script))
     # WORKAROUND for the fact that PDF.js tries to use the Fetch API even with
     # qute:// URLs.
-    pdfjs_script = '<script src="../build/pdf.js"></script>'
-    html = html.replace(pdfjs_script,
+    pdfjs_script = f'<script src="../build/{html.escape(pdfjs_name)}"></script>'
```
Comment (toofar) [inline qutebrowser/browser/pdfjs.py:67]: I don't think this works anymore because when using modules the source is now:

    <script src="../build/pdf.mjs" type="module"></script>
    
One way to fix it is to add a `{attrs}` variable in the f-string that is set to ` "type="module"` if the name ends with `.mjs`, otherwise it's empty.
An even easier way to fix it is to just put the workaround after `head` instead of after the script tag.

But I don't think this workaround is even active on the current pdfjs version anyhow. The code from here doesn't seem to be a thing anymore: https://github.com/qutebrowser/qutebrowser/issues/4235#issuecomment-423520775

**Comment by toofar** [comment]
> I'd look into possibly fixing our monkeypatching to stop PDF.js from using the Fetch API

Hmm, not sure we can. It doesn't seem to have the same fallback mechanism as described in https://github.com/qutebrowser/qutebrowser/issues/4235. Changing the URL it's loading from to be eg https://example.com/404 just leads to another error showing up. I don't think we can turn off this localization stuff unless we do some involved mocking of `this.externalServices` and `this.l10n`. We could maybe swap the URLs out for links to github...
Edit: see the last two paragraphs for more on this

> I'd rather not weaken permissions for qute: wholesale, as it will affect a lot.

I'm failing to hold in my head all the controls, or lack thereof, there are around this custom protocol. I don't know why fetch() requires additional controls compared to other ways of making requests. Normal CORS restrictions should apply here. We have [some stuff](https://github.com/qutebrowser/qutebrowser/blob/585c4fb/qutebrowser/browser/webengine/interceptor.py#L170) in the request interceptor (which we could probably expand, are there any qute:// URLs that should be called from a third party?). There's CSRF tokens for the settings page. Then there is all the upstream [flags](https://doc.qt.io/qt-6/qwebengineurlscheme.html#Flag-enum) for use when registering the scheme handler.

I don't understand what would be wrong with enabling the fetch API and ensuring the origin was restricted. In principle I mean, I'm not sure I would propose the refactoring of security controls required to make that convincing for this PR. 

I'm aware there was [this CVE](https://www.openwall.com/lists/oss-security/2018/07/11/7) about being able to set values despite CORS (why does it say I reported that!? I don't remember that.) I'm assuming that was because browser based CORS enforcing is only done against the response for some requests, so mutating requests can still have an effect on the server even if the response doesn't make it to the client. But we should be able to reject requests based on the Origin header just fine on the server side (eg the scheme handlers) before we've processed them (like we are doing in the interceptor for that one case).

> I think we should split it into qute-pdfjs:// etc.

Sure in this case, that sounds easier than having to navigate all the above mess while trying to keep in mind the other stuff on the qute:// scheme. In the longer term though I think if the security of the existing handlers was easier to understand we probably wouldn't be talking about adding a new scheme?

> and set fine-grained per-scheme permissions where needed.

What kind of fine grained permissions are you thinking of? Would checking `Origin.scheme() == qute-pdf` be enough? Or can malicious PDFs initiate arbitrary network requests? I haven't looked at all the pdf custom scheme handling but I think all it does is load files. So we should be able to ensure any paths are relative to the pdfjs installation, if we aren't already.

Ah, having a look at upstream it seems the fetch errors will be fixed in the next pdfjs version anyway. I think this fallback-less fetch was introduced in https://github.com/mozilla/pdf.js/pull/17115, which was in the latest release on November 6th.
It looks like that has already been changed in https://github.com/mozilla/pdf.js/pull/17277 to use a fetch wrapper that [is already aware](https://github.com/mozilla/pdf.js/blob/f4b396f6c88e951174879a52390fa89223326b36/src/display/display_utils.js#L805) of the limitations of the fetch API.

So it seems another option is to ship with these non-fatal fetch errors (and maybe suppress them? Although I think they do make some of the UI not work) and cut a new release after pdfjs does? We could even pull an older pdfjs release into the release builds if we wanted.

**Comment by toofar** [comment]
A new pdfjs release is out now that resolves the fetch() issues :tada: : https://github.com/mozilla/pdf.js/releases/tag/v4.0.269

Back to looking at the main point of this PR then.

**Comment by toofar** [comment]
I added a commit to fix :version detection, and a couple of minor commits. I haven't looked at tests beyond the version ones, or lint or anything.
That's me for the day.

**Comment by The-Compiler** [comment]
Quick update: v3.1.0 was released with this, and made its way to Archlinux.
